### ISSUE TITLE
Yum Module always displays as skipped 

### ISSUE THREAD

**Issue by juliedavila** [ansible/ansible-modules-core#1765]
##### Ansible Version:

```
ansible 2.0.0 (devel 919aaa5c42) last updated 2015/07/21 11:34:59 (GMT -400)
  lib/ansible/modules/core: (devel 8ce2a18921) last updated 2015/04/21 15:50:13 (GMT -400)
  lib/ansible/modules/extras: (devel db64a6b584) last updated 2015/04/21 15:50:19 (GMT -400)
```
##### Ansible Configuration:

Running with ANSIBLE_KEEP_REMOTE_FILES=1 and with pipelining enabled
##### Environment:

OSX targeting a CentOS 6.5 vagrant VM
##### Summary:

When using the yum module it always reports as skipped but still does the changes.
##### Steps To Reproduce:

``` yaml
  tasks:
    - yum:
        name: httpd
        state: absent
```
##### Expected Results:

``` shell
changed: [node-1] => {"changed": true, "msg": "", "rc": 0, "results": ["Loaded plugins: fastestmirror\nSetting up Remove Process\nResolving Dependencies\n--> Running transaction check\n---> Package httpd.x86_64 0:2.2.15-39.el6.centos will be erased\n--> Finished Dependency Resolution\n\nDependencies Resolved\n\n================================================================================\n Package       Arch           Version                       Repository     Size\n================================================================================\nRemoving:\n httpd         x86_64         2.2.15-39.el6.centos          @base         2.9 M\n\nTransaction Summary\n================================================================================\nRemove        1 Package(s)\n\nInstalled size: 2.9 M\nDownloading Packages:\nRunning rpm_check_debug\nRunning Transaction Test\nTransaction Test Succeeded\nRunning Transaction\n\r  Erasing    : httpd-2.2.15-39.el6.centos.x86_64                            1/1 \n\r  Verifying  : httpd-2.2.15-39.el6.centos.x86_64                            1/1 \n\nRemoved:\n  httpd.x86_64 0:2.2.15-39.el6.centos                                           \n\nComplete!\n"]}
```
##### Actual Results:

``` shell
skipping: [node-1] => {"changed": true, "msg": "", "rc": 0, "results": ["Loaded plugins: fastestmirror\nSetting up Remove Process\nResolving Dependencies\n--> Running transaction check\n---> Package httpd.x86_64 0:2.2.15-39.el6.centos will be erased\n--> Finished Dependency Resolution\n\nDependencies Resolved\n\n================================================================================\n Package       Arch           Version                       Repository     Size\n================================================================================\nRemoving:\n httpd         x86_64         2.2.15-39.el6.centos          @base         2.9 M\n\nTransaction Summary\n================================================================================\nRemove        1 Package(s)\n\nInstalled size: 2.9 M\nDownloading Packages:\nRunning rpm_check_debug\nRunning Transaction Test\nTransaction Test Succeeded\nRunning Transaction\n\r  Erasing    : httpd-2.2.15-39.el6.centos.x86_64                            1/1 \n\r  Verifying  : httpd-2.2.15-39.el6.centos.x86_64                            1/1 \n\nRemoved:\n  httpd.x86_64 0:2.2.15-39.el6.centos                                           \n\nComplete!\n"]}
```

Note that this occurs whether it actually removes/installs the package or on idempotent runs.

**Comment by Darwiner** [comment]
I'm also getting this issue and am currently on:

```
$ ansible --version
ansible 2.1.0 (devel adc9a25a06) last updated 2016/03/10 16:45:48 (GMT -400)
  lib/ansible/modules/core: (detached HEAD c86a0ef84a) last updated 2016/03/10 16:45:50 (GMT -400)
  lib/ansible/modules/extras: (detached HEAD 33a557cc59) last updated 2016/03/10 16:45:53 (GMT -400)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

Even though packages were updated (as can be seen in /var/log/yum.log), the task still shows as skipped:

```
TASK [server-management : HP - CentOS | Install packages] **********************
skipping: [ash-mc02.example.net] => (item=[u'cpqacuxe', u'hp-smh-templates', u'hpdiags', u'hp-snmp-agents', u'hpssa', u'hp-health', u'hponcfg', u'hpsmh'])
```

The task in question here is simply:

```
- name: HP - CentOS | Install packages
  yum: name={{ item }} enablerepo=HP-spp state=latest
  with_items:
    - cpqacuxe
    - hp-smh-templates
    - hpdiags
    - hp-snmp-agents
    - hpssa
    - hp-health
    - hponcfg
    - hpsmh
  when: ansible_distribution == 'CentOS'
  tags:
    - packages
```

**Comment by takuyakawabuchi** [comment]
I'm also getting this issue.

### Ansible Version:

```
ansible 2.1.0 (devel 3221be4340) last updated 2016/03/25 14:23:42 (GMT +900)
  lib/ansible/modules/core: (detached HEAD 7efc09ef08) last updated 2016/03/25 14:24:15 (GMT +900)
  lib/ansible/modules/extras: (detached HEAD 7f9cdc0350) last updated 2016/03/25 14:24:16 (GMT +900)
  config file = 
  configured module search path = Default w/o overrides
```

### Environment:

OSX Yosemite targeting a CentOS 6.7 vagrant VM

### Summary:

When using the yum module with with_items it always reports as skipped but still does the changes.

### Steps To Reproduce:

```
---
- hosts: all
  tasks:
    - yum:
        name: "{{ item.name }}"
        state: present
      become: yes
      with_items:
        - name: httpd
```

### Results:

```
TASK [yum] *********************************************************************
skipping: [centos67] => (item={u'name': u'httpd'})  => {"changed": true, "item": {"name": "httpd"}, "msg": "", "rc": 0, "results": ["Loaded plugins: fastestmirror, versionlock\nSetting up Install Process\nLoading mirror speeds from cached hostfile\n * base: mirror.readyspace.com\n * epel: mirror.wanxp.id\n * extras: mirror.vastspace.net\n * updates: mirror.0x.sg\nResolving Dependencies\n--> Running transaction check\n---> Package httpd.x86_64 0:2.2.15-47.el6.centos.4 will be installed\n--> Finished Dependency Resolution\n\nDependencies Resolved\n\n================================================================================\n Package      Arch          Version                        Repository      Size\n================================================================================\nInstalling:\n httpd        x86_64        2.2.15-47.el6.centos.4         updates        831 k\n\nTransaction Summary\n================================================================================\nInstall       1 Package(s)\n\nTotal download size: 831 k\nInstalled size: 2.9 M\nDownloading Packages:\nRunning rpm_check_debug\nRunning Transaction Test\nTransaction Test Succeeded\nRunning Transaction\n\r  Installing : httpd-2.2.15-47.el6.centos.4.x86_64                          1/1 \n\r  Verifying  : httpd-2.2.15-47.el6.centos.4.x86_64                          1/1 \n\nInstalled:\n  httpd.x86_64 0:2.2.15-47.el6.centos.4                                         \n\nComplete!\n"]}
```

**Comment by barney3000** [comment]
Same problem here.

## Ansible Version:

`ansible 2.1.0`
`config file = /etc/ansible/ansible.cfg`
`configured module search path = Default w/o overrides`

## Environment:

CentOS 7.2.1511 targeting a Red Hat EL 6.7 (Santiago) on VMWare vm.

## Summary:

When using the yum module with with_items it always reports as skipped but still does the changes.
Example of the steps followed on the roles:

`---`
`name: Install packages`
`yum: name={{ item.pkgs_core }} state=installed`
`with_items: "{{ pkgs_list }}"`

`pkgs_list:`
`- { pkgs_core: curl }`
`- { pkgs_core: libcurl-devel }`
`- { pkgs_core: openssl }`

## Results:

`skipping: [eidosdev] => (item={u'pkgs_core': u'curl'})`
`skipping: [eidosdev] => (item={u'pkgs_core': u'libcurl-devel'})`
`skipping: [eidosdev] => (item={u'pkgs_core': u'openssl'})`

Even if it says skipping, the packages has been actually installed.

Thank you,

**Comment by gurumaia** [comment]
Same here.

# Version

`ansible 2.1.0 (devel 22467a0de8) last updated 2016/04/13 10:53:15 (GMT -200)
  lib/ansible/modules/core: (detached HEAD 99cd31140d) last updated 2016/04/13 10:53:53 (GMT -200)
  lib/ansible/modules/extras: (detached HEAD ab2f4c4002) last updated 2016/04/13 10:54:38 (GMT -200)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides`

# Output

`skipping: [ec2_site_hml_hiril] => (item={u'version': u'1.3.6-5.el6', u'name': u'npm'})  => {"changed": false, "invocation": {"module_args": {"conf_file": null, "disable_gpg_check": false, "disablerepo": null, "enablerepo": null, "exclude": null, "install_repoquery": true, "list": null, "name": ["npm-1.3.6-5.el6"], "state": "present", "update_cache": false, "validate_certs": true}, "module_name": "yum"}, "item": {"name": "npm", "version": "1.3.6-5.el6"}, "msg": "", "rc": 0, "results": ["npm-1.3.6-5.el6.noarch providing npm-1.3.6-5.el6 is already installed"]}`
So it states that it's already installed, meaning it tried to install it without checking.

**Comment by orenwolf** [comment]
Confirmed as well, against RHEL 7.2 target from an OSX host via homebrew:

```
$ ansible --version
ansible 2.0.2.0
```

**Comment by vaurdan** [comment]
Confirming this too on 2.1.0, running against CentOS 7 running from a CentOS 7 host:

```
$ ansible --version
ansible 2.1.0
```

**Comment by percygrunwald** [comment]
Confirmed against CentOS 6 & 7 targets from an OS X host via homebrew:

``` bash
$ ansible --version
ansible 2.0.2.0
```

**Comment by ghost** [comment]
Another confirmation. CentOS 7.2.151

`ansible 2.0.2.0`

We're dead in the water now.

**Comment by smalenfant** [comment]
This was seen also as part of #3574. Is this bug almost a year old?

**Comment by nitzmahone** [comment]
Should be fixed in 2.1 RC2 with https://github.com/ansible/ansible/commit/133395db30110175b02325a23ff5156077f71458

**Comment by rsumner** [comment]
As of 2.1 RC4, this is still not rolled up.

**Comment by nitzmahone** [comment]
@rsumner: Hmm, the bug in question appears fixed for me in 2.1RC4. 

```
- hosts: linux
  gather_facts: no
  tasks:
  - yum:
      name: "{{ item }}"
    with_items:
    - perl
    - git
```

shows "changed" the first time through and "ok" the second (it used to show "skipped", which was the original bug).

**Comment by rsumner** [comment]
Doh, wasn't seeing your patch due to a lame RPM spec.  A rebuild with the correct source shows a change.  My problem still exists that if one of the specified with_items isn't available, the whole set will fail.  Is this still behaving as designed?  On the contrary, if I run a "yum install" consisting of a set of packages with one not being available, the rest of the packages will still install.

**Comment by smalenfant** [comment]
It seems like proper behavior to me. Try putting in ignore_errors and see
if that works. Should provide behavior you are looking for.

**Comment by nitzmahone** [comment]
Yes, definitely "as designed" - defaulting to silent error behavior would
be somewhat disastrous. You can opt in if you want that behavior
(ignore_errors or sniff output with failed_when).

**Comment by rsumner** [comment]
ignore_errors doesn't change the result.  output of my playbook with -vvv:

`{"changed": false, "failed": true, "invocation": {"module_args": {"conf_file": null, "disable_gpg_check": false, "disablerepo": null, "enablerepo": null, "exclude": null, "install_repoquery": true, "list": null, "name": ["busybox", "kbd-misc", "dbus-python", "dstat", "newt-python", "sssd-proxy", "mpfr"], "msg": "No Package matching 'busybox' found available, installed or updated", "rc": 0, "results": []}`

and my playbook:

`- name: install packages
    yum: name={{ item }} state=latest
    with_items: "{{ hostvars[other_server]['rpm_list'].keys() }}"
    ignore_errors: yes`

**Comment by rsumner** [comment]
If i run the same via "yum install" supplying all of these packages, it will, of course complain of "No package busybox available", however it will allow me to proceed with installation of the remaining packages (which do exist).  I'm hoping for the same functionality here, however I'm willing to resort to using "command" if needed.

**Comment by smalenfant** [comment]
I think I know what you mean... You expect the packages to be installed
even if one fails, but then ignore_errors should make it continue. In this
case here, mtr doesn't get installed because tata is invalid.

Seems like this should be into a new issue, although been working like this
for a while I guess.

For example using Ansible 1.9.6...

TASK: [install packages]

---

failed: [cdn1cdelas0002] => (item=tata,mtr) => {"changed": false, "failed":
true, "item": "tata,mtr", "rc": 0, "results": []}
msg: No Package matching 'tata' found available, installed or updated
...ignoring

PLAY RECAP

---

cdn1cdelas0002 : ok=2    changed=0    unreachable=0    failed=0

**Comment by rsumner** [comment]
@smalenfant you got it.  Sounds like this should be a new issue, but at least with_items are working like they did <2.1 RC1 with nitzmahone's fix.

**Comment by rsumner** [comment]
Heh, and line 827 of yum.py:

`827             # FIXME..? may not be desirable to throw an exception here if a single package is missing`

**Comment by gamename** [comment]
Agree.  This should probably be a new issue.  I'm seeing the behavior on 2.0.2.0

**Comment by ghost** [comment]
Same here... Ansible 2.0.2.0 on CentOS 6 and 7 (From epel-testing):

```
    qemu: TASK [bootstrap : centos6 - yum installing basic tools] ************************
    qemu: skipping: [127.0.0.1] => (item=[u'at', u'nc', u'git', u'vim', u'man', u'zip', u'ntp', u'gpm', u'lftp', u'curl', u'wget', u'lsof', u'tmux', u'nmap', u'unzip', u'acpid', u'iotop', u'iftop', u'which', u'figlet', u'iperf3', u'bwm-ng', u'telnet', u'parted', u'psmisc', u'cronie', u'tcpick', u'sysstat', u'rsyslog', u'tcpdump', u'numactl', u'mlocate', u'ansible', u'pciutils', u'deltarpm', u'dos2unix', u'dmidecode', u'man-pages', u'tcpreplay', u'bind-utils', u'traceroute', u'conntrack-tools', u'libhugetlbfs-utils']) 
```

But, it is being applied!

**Comment by nitzmahone** [comment]
@tmartinx this bug is fixed in 2.1
